src/transaction/atomic_replication_helper.cpp

```diff
@@ -98,24 +96,60 @@ namespace cublog
 	return;
       }
 
-    for (size_t i = 0; i < iterator->second.size (); i++)
-      {
-	iterator->second[i].unfix_page (thread_p);
-      }
-    iterator->second.clear ();
+    iterator->second.unfix_sequence (thread_p);
     m_atomic_sequences_map.erase (iterator);
 
 #if !defined (NDEBUG)
     m_atomic_sequences_vpids_map.erase (tranid);
 #endif
   }
+  /****************************************************************************
+   * atomic_replication_helper::atomic_replication_unit function definitions  *
+   ****************************************************************************/
+  template <typename T>
+  int atomic_replication_helper::atomic_replication_sequence::add_atomic_replication_unit (THREAD_ENTRY *thread_p,
+      log_lsa record_lsa, LOG_RCVINDEX rcvindex, VPID vpid, log_rv_redo_context &redo_context,
+      const log_rv_redo_rec_info<T> &record_info)
+  {
+    m_atomic_replication_unit_vector.emplace_back (record_lsa, vpid, rcvindex);
+    auto iterator = m_atomic_sequence_pages_map.find (vpid);
+    if (iterator == m_atomic_sequence_pages_map.cend ())
+      {
+	int error_code = m_atomic_replication_unit_vector.back ().fix_page (thread_p);
+	if (error_code != NO_ERROR)
+	  {
+	    return error_code;
+	  }
+	m_atomic_sequence_pages_map.emplace (vpid,  m_atomic_replication_unit_vector.back ().get_page_ptr ());
+      }
+    else
+      {
+	m_atomic_replication_unit_vector.back ().set_page_ptr (iterator->second ());
+      }
+    m_atomic_replication_unit_vector.back ().apply_log_redo (thread_p, redo_context, record_info);
```

Comment:
I don't know if this is the final version of the code or not:
but having both page fix and apply redo in the same phase may lead to undesired page unfixes of previously fixed pages during current ordered fix.
If a future refactoring task will separate into two phases, please ignore this.

Reply:
It will not remain like this. It will be a separate modification - http://jira.cubrid.org/browse/LETS-460 .